src/main/java/org/terasology/web/serverAdminManagement/AdminPermissionManager.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2018 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.web.serverAdminManagement;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.engine.paths.PathManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.logic.permission.PermissionManager;
+import org.terasology.logic.permission.PermissionSetComponent;
+import org.terasology.network.ClientComponent;
+import org.terasology.web.resources.base.ResourceMethodName;
+import org.terasology.web.resources.base.ResourcePath;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.lang.reflect.Type;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Manages the serverAdminPermissions.json file in the server directory.
+ */
+public final class AdminPermissionManager extends BaseComponentSystem {
+
+    private static final Logger logger = LoggerFactory.getLogger(AdminPermissionManager.class);
+    private static final Gson GSON = new Gson();
+    private static AdminPermissionManager instance;
+
+    private final Path adminPermissionsFilePath;
+    private final Type typeOfServerAdminPermissions = new TypeToken<Set<AdminPermissions>>() { }.getType();
+    private Set<AdminPermissions> serverAdminPermissions;
+    private Runnable onListChanged = () -> { };
+
+    private AdminPermissionManager(Path adminPermissionsFilePath) {
+        this.adminPermissionsFilePath = adminPermissionsFilePath;
+        setServerAdminPermissions(new HashSet<>());
+    }
+
+    public static AdminPermissionManager getInstance() {
+        if (instance == null) {
+            instance = new AdminPermissionManager(PathManager.getInstance().getHomePath().resolve("serverAdminPermissions.json"));
+        }
+        return instance;
+    }
+
+    public boolean adminHasPermission(String adminID, ResourcePath path, ResourceMethodName resourceMethodName) {
```

Comment:
I think a better implementation could be if the resources themselves - or better, each method of each resource - knew which permissions are required to access them. For example, the way I checked for permissions (which, as you already know, were just binary before your code - a client was either an admin or a non-admin, there were no permission levels) made use of the `ClientSecurityRequirements` class. An instance of this class is created for each resource method and passed to the method builders (in `ResourceMethodFactory`, like `createVoidParameterlessMethod` and similar), and encapsulates what are the requirements to execute a certain method (`GET`, `POST`, etc) on a certain resource. Maybe you could modify `ClientSecurityRequirements` to also encapsulate which permission type is required, and have its `clientIsAllowed` method somehow query `AdminPermissionManager` to determine its result?